File: tests/property/python/runner.py
#!/usr/bin/env sage
"""
Property test runner for SageMath.
Executes operations with deterministic seeding and outputs results.

This script reads test case definitions from stdin (JSON format) and
executes them in SageMath, outputting results in a format that can be
compared with the TypeScript implementation.

Usage:
    sage runner.py < test_cases.json
    sage runner.py --seed 42 < test_cases.json

Output format:
    {"function": "gcd", "args": [12, 8], "result": "4", "seed": 42}

Areas
-----
This runner contains NO area-specific code.  Every area lives in its own module
under ``areas/`` and is discovered by globbing that directory, so several agents
can add areas in parallel without ever editing this file.  See
``tests/property/README.md``.
"""

import glob
import importlib
import json
import os
import sys
import random as py_random
from sage.all import *

# Importing areas/ would otherwise leave a __pycache__ directory inside the repo
# (which .gitignore does not cover); the runner is short-lived so caching buys
# nothing.
sys.dont_write_bytecode = True

# Make ``areas`` importable as a package no matter which directory `sage` was
# invoked from (compare.ts runs us with cwd = project root).
_RUNNER_DIR = os.path.dirname(os.path.abspath(__file__))
if _RUNNER_DIR not in sys.path:
    sys.path.insert(0, _RUNNER_DIR)

AREAS_DIR = os.path.join(_RUNNER_DIR, 'areas')

# name -> loaded {'functions': ..., 'formatters': ...}; filled in on demand.
_LOADED_AREAS = {}


def discover_area_names():
    """
    Return the sorted names of the area modules present in ``areas/``.

    An area is any ``areas/<area>.py`` whose basename does not start with ``_``
    (underscore-prefixed modules are shared helpers, not areas).  The name must
    match the ``module`` field of ``cases/<area>.cases.json``.
    """
    names = []
    for path in sorted(glob.glob(os.path.join(AREAS_DIR, '*.py'))):
        name = os.path.splitext(os.path.basename(path))[0]
        if name.startswith('_'):
            continue
        names.append(name)
    return names


def load_area(name):
    """
    Import ``areas/<name>.py`` and return its dispatch tables.

    Every area module must export a ``FUNCTIONS`` dict mapping the ``function``
    names used in ``cases/<name>.cases.json`` to callables.  It may optionally
    export a ``FORMATTERS`` dict mapping a function name to a ``result -> str``
    callable, for results the generic ``format_result`` below cannot render.

    Areas are imported lazily and one at a time, so a broken area module can
    never take down the areas owned by other agents.

    Returns:
        dict: {'functions': dict, 'formatters': dict}
    """
    if name in _LOADED_AREAS:
        return _LOADED_AREAS[name]

    module = importlib.import_module('areas.' + name)
    functions = getattr(module, 'FUNCTIONS', None)
    if not isinstance(functions, dict):
        raise ValueError(
            f"Area module areas/{name}.py must export a FUNCTIONS dict "
            f"mapping function names to callables"
        )
    formatters = getattr(module, 'FORMATTERS', {})
    _LOADED_AREAS[name] = {'functions': functions, 'formatters': formatters}
    return _LOADED_AREAS[name]


def set_seed(seed):
    """Set random seed for reproducibility across both SageMath and Python random."""
    set_random_seed(seed)
    py_random.seed(seed)


def generate_arg(generator_spec, seed):
    """
    Generate an argument based on the generator specification.

    Supported generators:
    - randomBigint(min, max): Random integer in range [min, max]
    - randomPrime(min, max): Random prime in range [min, max]
    - fixedValue(value): Fixed value
    - randomList(generator, length): List of random values

    Note: Uses Python's random module (Mersenne Twister) to match TypeScript.
    """
    if generator_spec.startswith('randomBigint('):
        # Parse: randomBigint(min, max)
        params = generator_spec[13:-1]  # Remove 'randomBigint(' and ')'
        parts = params.split(',')
        min_val = int(parts[0].strip())
        max_val = int(parts[1].strip())
        # Use Python's random.randint to match TypeScript's Mersenne Twister
        return Integer(py_random.randint(min_val, max_val))

    elif generator_spec.startswith('randomPrime('):
        # Parse: randomPrime(min, max)
        params = generator_spec[12:-1]
        parts = params.split(',')
        min_val = int(parts[0].strip())
        max_val = int(parts[1].strip())
        # Use Python's random with rejection sampling to match TypeScript
        max_attempts = 10000
        for _ in range(max_attempts):
            candidate = py_random.randint(min_val, max_val)
            if is_prime(candidate):
                return Integer(candidate)
        # Fallback: find next prime from min
        p = next_prime(min_val - 1)
        return p

    elif generator_spec.startswith('fixedValue('):
        # Parse: fixedValue(value) - value can be integer or array
        value_str = generator_spec[11:-1].strip()
        if value_str.startswith('[') and value_str.endswith(']'):
            # Parse array: [1, 2, 3]
            inner = value_str[1:-1].strip()
            if not inner:
                return []
            parts = inner.split(',')
            return [Integer(p.strip()) for p in parts]
        return Integer(value_str)

    elif generator_spec.startswith('randomList('):
        # Parse: randomList(generator, length)
        params = generator_spec[11:-1]
        # Find the last comma that separates length
        last_comma = params.rfind(',')
        inner_generator = params[:last_comma].strip()
        length = int(params[last_comma+1:].strip())
        return [generate_arg(inner_generator, seed + i) for i in range(length)]

    else:
        # Assume it's a literal value
        return Integer(generator_spec)


def format_result(result):
    """
    Format a SageMath result as a string for comparison.

    This ensures consistent string representation across Python and TypeScript.
    """
    if result is None:
        return 'null'
    if isinstance(result, bool):
        return 'True' if result else 'False'
    elif isinstance(result, Factorization):
        # Format: "2^2 * 3" or "1" for empty
        if len(result) == 0:
            return '1'
        parts = []
        for (p, e) in result:
            if e == 1:
                parts.append(str(p))
            else:
                parts.append(f'{p}^{e}')
        return ' * '.join(parts)
    elif isinstance(result, (list, tuple)):
        # Format lists/tuples to match Python's str() output
        if isinstance(result, tuple):
            return '(' + ', '.join(format_result(x) for x in result) + ')'
        else:
            return '[' + ', '.join(format_result(x) for x in result) + ']'
    elif hasattr(result, '__iter__') and not isinstance(result, str):
        # Other iterables
        return '[' + ', '.join(format_result(x) for x in result) + ']'
    else:
        return str(result)


def execute_function(module, function_name, args):
    """
    Execute a SageMath function with the given arguments.

    Args:
        module: Area name (e.g., "arith", "elliptic_curves"), matching areas/<module>.py
        function_name: Function name (e.g., "gcd", "is_prime")
        args: List of arguments

    Returns:
        The result of the function call
    """
    if module not in discover_area_names():
        raise ValueError(f"Unknown module: {module}")

    functions = load_area(module)['functions']
    if function_name not in functions:
        raise ValueError(f"Unknown function: {module}.{function_name}")

    func = functions[function_name]
    return func(*args)


def format_for(module, function_name, result):
    """Format a result, honouring an area-specific FORMATTERS override if present."""
    if module not in discover_area_names():
        return format_result(result)
    formatter = load_area(module)['formatters'].get(function_name)
    if formatter is not None:
        return formatter(result)
    return format_result(result)


def run_test_case(case, seed):
    """
    Run a single test case with the given seed.

    Args:
        case: Test case definition dict
        seed: Random seed to use

    Returns:
        Result dict with function, args, result, and seed
    """
    set_seed(seed)

    module = case.get('module', 'arith')
    function_name = case['function']
    arg_generators = case.get('argGenerators', [])

    # Generate arguments
    args = []
    for i, gen in enumerate(arg_generators):
        arg = generate_arg(gen, seed + i * 1000)
        args.append(arg)

    # Execute function
    try:
        result = execute_function(module, function_name, args)
        formatted_result = format_for(module, function_name, result)
        error = None
    except Exception as e:
        formatted_result = None
        error = str(e)

    return {
        'function': function_name,
        'args': [str(a) for a in args],
        'result': formatted_result,
        'error': error,
        'seed': seed,
    }


def run_test_suite(test_suite):
    """
    Run all test cases in a test suite.

    Args:
        test_suite: Dict with 'module' and 'cases' keys

    Returns:
        List of result dicts
    """
    module = test_suite.get('module', 'arith')
    cases = test_suite.get('cases', [])
    results = []

    for case in cases:
        case['module'] = module
        seeds = case.get('seeds', [42])

        for seed in seeds:
            result = run_test_case(case, seed)
            results.append(result)

    return results


def main():
    """Main entry point."""
    # Parse command line arguments
    seed_override = None
    for i, arg in enumerate(sys.argv[1:], 1):
        if arg == '--seed' and i < len(sys.argv) - 1:
            seed_override = int(sys.argv[i + 1])

    # Read test cases from stdin
    input_data = sys.stdin.read()

    if not input_data.strip():
        print("Error: No input provided. Pass test cases as JSON via stdin.", file=sys.stderr)
        sys.exit(1)

    try:
        test_suite = json.loads(input_data)
    except json.JSONDecodeError as e:
        print(f"Error parsing JSON: {e}", file=sys.stderr)
        sys.exit(1)

    # Override seeds if specified
    if seed_override is not None:
        for case in test_suite.get('cases', []):
            case['seeds'] = [seed_override]

    # Run tests
    results = run_test_suite(test_suite)

    # Output results as JSON
    print(json.dumps(results, indent=2))


if __name__ == '__main__':
    main()

File: tests/property/typescript/runner.ts
#!/usr/bin/env bun
/**
 * Property test runner for sagemath-ts.
 * Executes operations with deterministic seeding and outputs results.
 *
 * This script reads test case definitions from stdin (JSON format) and
 * executes them using our TypeScript implementation, outputting results
 * in a format that can be compared with the SageMath implementation.
 *
 * Usage:
 *   bun runner.ts < test_cases.json
 *   bun runner.ts --seed 42 < test_cases.json
 *
 * Output format:
 *   {"function": "gcd", "args": [12, 8], "result": "4", "seed": 42}
 *
 * Areas
 * -----
 * This runner contains NO area-specific code. Every area lives in its own module
 * under `areas/` and is discovered by globbing that directory, so several agents
 * can add areas in parallel without ever editing this file. See
 * tests/property/README.md.
 */

import { existsSync, readdirSync } from 'node:fs';
import { join } from 'node:path';
import { pathToFileURL } from 'node:url';
import {
  type Factorization,
  formatFactorization,
  is_prime,
  next_prime,
} from '../../../packages/sagemath-ts/src/index.js';
import { MersenneTwister } from './mersenne-twister.js';

/** Directory holding the per-area dispatch modules. */
const AREAS_DIR = join(import.meta.dir, 'areas');

/**
 * Seeded pseudo-random number generator using Mersenne Twister.
 * This matches Python's random module exactly for the same seed.
 */
class SeededRandom {
  private mt: MersenneTwister;

  constructor(seed: number) {
    this.mt = new MersenneTwister(seed);
  }

  /**
   * Generate a random integer in range [min, max] inclusive
   */
  integer(min: bigint, max: bigint): bigint {
    return this.mt.randint(min, max);
  }

  /**
   * Generate a random prime in range [min, max]
   */
  prime(min: bigint, max: bigint): bigint {
    // Simple approach: generate random numbers until we find a prime
    // This is not efficient but works for testing with small ranges
    const maxAttempts = 10000;
    for (let i = 0; i < maxAttempts; i++) {
      const candidate = this.integer(min, max);
      if (is_prime(candidate)) {
        return candidate;
      }
    }
    // Fallback: find next prime from min
    const p = next_prime(min - 1n);
    while (p > max) {
      // If no prime in range, just return the next prime after min
      return p;
    }
    return p;
  }
}

/**
 * Test case definition from JSON
 */
interface TestCase {
  module?: string;
  function: string;
  seeds?: number[];
  argGenerators?: string[];
}

/**
 * Test suite definition from JSON
 */
interface TestSuite {
  module: string;
  cases: TestCase[];
}

/**
 * Test result output
 */
interface TestResult {
  function: string;
  args: string[];
  result: string | null;
  error: string | null;
  seed: number;
}

/**
 * A discovered area: its dispatch table plus any per-function formatter overrides.
 */
interface Area {
  functions: Record<string, (...args: never[]) => unknown>;
  formatters: Record<string, (result: unknown) => string>;
}

/**
 * Return the names of the area modules present in `areas/`.
 *
 * An area is any `areas/<area>.ts` whose basename does not start with `_`
 * (underscore-prefixed modules are shared helpers, not areas). The name must
 * match the `module` field of `cases/<area>.cases.json`.
 */
function discoverAreaNames(): string[] {
  if (!existsSync(AREAS_DIR)) {
    return [];
  }
  return readdirSync(AREAS_DIR)
    .filter(
      (f) =>
        f.endsWith('.ts') && !f.startsWith('_') && !f.endsWith('.d.ts') && !f.endsWith('.test.ts')
    )
    .map((f) => f.slice(0, -'.ts'.length))
    .sort();
}

/**
 * Import `areas/<name>.ts` and return its dispatch tables.
 *
 * Every area module must export a `functions` object mapping the `function`
 * names used in `cases/<name>.cases.json` to callables. It may optionally
 * export a `formatters` object mapping a function name to a
 * `(result) => string` callable, for results the generic `formatResult` below
 * cannot render.
 *
 * Areas are imported lazily and one at a time, so a broken area module can
 * never take down the areas owned by other agents.
 */
async function loadArea(name: string): Promise<Area> {
  const url = pathToFileURL(join(AREAS_DIR, `${name}.ts`)).href;
  const mod = (await import(url)) as {
    functions?: unknown;
    formatters?: unknown;
  };

  if (typeof mod.functions !== 'object' || mod.functions === null) {
    throw new Error(
      `Area module areas/${name}.ts must export a \`functions\` object mapping function names to callables`
    );
  }

  return {
    functions: mod.functions as Record<string, (...args: never[]) => unknown>,
    formatters: (mod.formatters ?? {}) as Record<string, (result: unknown) => string>,
  };
}

/**
 * Generate an argument based on the generator specification.
 */
function generateArg(generatorSpec: string, random: SeededRandom): bigint | bigint[] {
  if (generatorSpec.startsWith('randomBigint(')) {
    // Parse: randomBigint(min, max)
    const params = generatorSpec.slice(13, -1);
    const parts = params.split(',');
    const minVal = BigInt(parts[0]!.trim());
    const maxVal = BigInt(parts[1]!.trim());
    return random.integer(minVal, maxVal);
  }

  if (generatorSpec.startsWith('randomPrime(')) {
    // Parse: randomPrime(min, max)
    const params = generatorSpec.slice(12, -1);
    const parts = params.split(',');
    const minVal = BigInt(parts[0]!.trim());
    const maxVal = BigInt(parts[1]!.trim());
    return random.prime(minVal, maxVal);
  }

  if (generatorSpec.startsWith('fixedValue(')) {
    // Parse: fixedValue(value) - value can be integer or array
    const valueStr = generatorSpec.slice(11, -1).trim();
    if (valueStr.startsWith('[') && valueStr.endsWith(']')) {
      // Parse array: [1, 2, 3]
      const inner = valueStr.slice(1, -1).trim();
      if (!inner) {
        return [];
      }
      const parts = inner.split(',');
      return parts.map((p) => BigInt(p.trim()));
    }
    return BigInt(valueStr);
  }

  if (generatorSpec.startsWith('randomList(')) {
    // Parse: randomList(generator, length)
    const params = generatorSpec.slice(11, -1);
    const lastComma = params.lastIndexOf(',');
    const innerGenerator = params.slice(0, lastComma).trim();
    const length = Number.parseInt(params.slice(lastComma + 1).trim());
    const result: bigint[] = [];
    for (let i = 0; i < length; i++) {
      const val = generateArg(innerGenerator, random);
      if (Array.isArray(val)) {
        result.push(...val);
      } else {
        result.push(val);
      }
    }
    return result;
  }

  // Assume it's a literal value
  return BigInt(generatorSpec);
}

/**
 * Format a result to match SageMath's string representation.
 *
 * This is the generic fallback used by every area. If your area produces a
 * shape this cannot render, either return an already-formatted string from the
 * area function or export a `formatters` entry from the area module — do not
 * add another special case here (it would make this file a merge hotspot).
 */
function formatResult(result: unknown, functionName?: string): string {
  // Handle null
  if (result === null) {
    return 'null';
  }

  // Special case: factor always returns a Factorization
  if (functionName === 'factor' && Array.isArray(result)) {
    return formatFactorization(result as Factorization);
  }

  // Special case: poly_factor returns [(coeffs, mult), ...]
  if (
    (functionName === 'poly_factor_ff' || functionName === 'poly_factor') &&
    Array.isArray(result)
  ) {
    // Format polynomial factorization as list of (factor_coeffs, multiplicity) pairs
    const formatted = result.map((item) => {
      if (Array.isArray(item) && item.length === 2 && Array.isArray(item[0])) {
        const [coeffs, mult] = item as [bigint[], number];
        const coeffsStr = '[' + coeffs.map((c) => c.toString()).join(', ') + ']';
        return `(${coeffsStr}, ${mult})`;
      }
      return formatResult(item);
    });
    return '[' + formatted.join(', ') + ']';
  }

  // Special case: poly_quo_rem returns (quotient, remainder) as tuple
  if (
    (functionName === 'poly_quo_rem_ff' || functionName === 'poly_quo_rem') &&
    Array.isArray(result)
  ) {
    if (result.length === 2 && Array.isArray(result[0]) && Array.isArray(result[1])) {
      const [q, r] = result as [bigint[], bigint[]];
      const qStr = '[' + q.map((c) => c.toString()).join(', ') + ']';
      const rStr = '[' + r.map((c) => c.toString()).join(', ') + ']';
      return `(${qStr}, ${rStr})`;
    }
  }

  if (typeof result === 'boolean') {
    return result ? 'True' : 'False';
  }

  if (Array.isArray(result)) {
    // Check if it's a 2D array (matrix) for HNF/LLL results
    if (
      result.length > 0 &&
      Array.isArray(result[0]) &&
      result[0].length > 0 &&
      (functionName?.startsWith('hnf_') || functionName?.startsWith('lll_'))
    ) {
      // It's a matrix result - format as list of lists
      const formattedRows = result.map(
        (row: Array<unknown>) => '[' + row.map((x) => String(x)).join(', ') + ']'
      );
      return '[' + formattedRows.join(', ') + ']';
    }

    // Check if it looks like a factorization (array of [prime, exp] tuples)
    // Must have length 2 inner arrays where both elements are bigint
    // But exclude matrix operations that return 2D arrays
    if (
      result.length > 0 &&
      Array.isArray(result[0]) &&
      result[0].length === 2 &&
      typeof result[0][0] === 'bigint' &&
      typeof result[0][1] === 'bigint' &&
      // Exclude matrix functions
      !functionName?.startsWith('hnf_') &&
      !functionName?.startsWith('lll_') &&
      !functionName?.startsWith('snf_')
    ) {
      // It's a Factorization
      return formatFactorization(result as Factorization);
    }

    // Check if it's a tuple-like result from xgcd (always 3 elements)
    // Only format as tuple if this is specifically xgcd function
    if (
      functionName === 'xgcd' &&
      result.length === 3 &&
      result.every((x) => typeof x === 'bigint')
    ) {
      return `(${result.map((x) => x!.toString()).join(', ')})`;
    }

    // Check if it's a signature result (tuple of 2 numbers)
    if (
      functionName === 'quadratic_signature' &&
      result.length === 2 &&
      result.every((x) => typeof x === 'number')
    ) {
      return `(${result[0]}, ${result[1]})`;
    }

    // Regular array
    return '[' + result.map((x) => formatResult(x)).join(', ') + ']';
  }

  if (typeof result === 'bigint') {
    return result.toString();
  }

  return String(result);
}

/**
 * Execute a function with the given arguments.
 */
function executeFunction(
  module: string,
  area: Area | undefined,
  functionName: string,
  args: (bigint | bigint[])[]
): unknown {
  if (area === undefined) {
    throw new Error(`Unknown module: ${module}`);
  }

  if (!(functionName in area.functions)) {
    throw new Error(`Unknown function: ${module}.${functionName}`);
  }

  const func = area.functions[functionName]! as (...a: unknown[]) => unknown;
  return func(...args);
}

/**
 * Run a single test case with the given seed.
 */
function runTestCase(
  testCase: TestCase,
  seed: number,
  module: string,
  area: Area | undefined
): TestResult {
  const random = new SeededRandom(seed);
  const functionName = testCase.function;
  const argGenerators = testCase.argGenerators || [];

  // Generate arguments
  const args: (bigint | bigint[])[] = [];
  for (const gen of argGenerators) {
    const arg = generateArg(gen, random);
    args.push(arg);
  }

  // Execute function
  let formattedResult: string | null = null;
  let error: string | null = null;

  try {
    const result = executeFunction(module, area, functionName, args);
    const override = area?.formatters[functionName];
    formattedResult = override ? override(result) : formatResult(result, functionName);
  } catch (e) {
    error = e instanceof Error ? e.message : String(e);
  }

  return {
    function: functionName,
    args: args.map((a) => (Array.isArray(a) ? `[${a.join(', ')}]` : a.toString())),
    result: formattedResult,
    error,
    seed,
  };
}

/**
 * Run all test cases in a test suite.
 */
async function runTestSuite(testSuite: TestSuite): Promise<TestResult[]> {
  const module = testSuite.module || 'arith';
  const cases = testSuite.cases || [];
  const results: TestResult[] = [];

  // Load only the area this suite needs, so a broken sibling area module can
  // never break an unrelated suite.
  const area = discoverAreaNames().includes(module) ? await loadArea(module) : undefined;

  for (const testCase of cases) {
    const seeds = testCase.seeds || [42];

    for (const seed of seeds) {
      const result = runTestCase(testCase, seed, module, area);
      results.push(result);
    }
  }

  return results;
}

/**
 * Main entry point
 */
async function main() {
  // Parse command line arguments
  let seedOverride: number | null = null;
  const args = process.argv.slice(2);

  for (let i = 0; i < args.length; i++) {
    if (args[i] === '--seed' && i + 1 < args.length) {
      seedOverride = Number.parseInt(args[i + 1]!);
      break;
    }
  }

  // Read test cases from stdin
  const inputChunks: string[] = [];
  const reader = Bun.stdin.stream().getReader();

  while (true) {
    const { done, value } = await reader.read();
    if (done) break;
    inputChunks.push(new TextDecoder().decode(value));
  }

  const inputData = inputChunks.join('');

  if (!inputData.trim()) {
    console.error('Error: No input provided. Pass test cases as JSON via stdin.');
    process.exit(1);
  }

  let testSuite: TestSuite;
  try {
    testSuite = JSON.parse(inputData);
  } catch (e) {
    console.error(`Error parsing JSON: ${e}`);
    process.exit(1);
  }

  // Override seeds if specified
  if (seedOverride !== null) {
    for (const testCase of testSuite.cases || []) {
      testCase.seeds = [seedOverride];
    }
  }

  // Run tests
  const results = await runTestSuite(testSuite);

  // Output results as JSON
  console.log(JSON.stringify(results, null, 2));
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});
